src/internal/connector/onedrive/restore.go

```diff
@@ -696,26 +797,55 @@ func restoreData(
 		return "", details.ItemInfo{}, err
 	}
 
-	// Get a drive item writer
-	w, err := driveItemWriter(ctx, service, driveID, ptr.Val(newItem.GetId()), ss.Size())
-	if err != nil {
-		return "", details.ItemInfo{}, err
-	}
+	var (
+		w       io.Writer
+		written int64
+	)
 
-	iReader := itemData.ToReader()
-	progReader, closer := observe.ItemProgress(
-		ctx,
-		iReader,
-		observe.ItemRestoreMsg,
-		clues.Hide(name),
-		ss.Size())
+	// This is just to retry file upload, the uploadSession creation is not retried here
+	// We need extra logic to retry file upload as we have to pull the file again from kopia
+	for i := 0; i <= maxUploadRetries; i++ {
+		// Get a drive item writer
+		w, err = driveItemWriter(ctx, service, driveID, ptr.Val(newItem.GetId()), ss.Size())
+		if err != nil {
+			return "", details.ItemInfo{}, err
+		}
+
+		pname := name
+		iReader := itemData.ToReader()
+
+		if i > 0 {
+			pname = fmt.Sprintf("%s (retry %d)", name, i)
+
+			// If it is not the first try, we have to pull the file
+			// again from kopia. Ideally we could just seek the stream
+			// but we don't have a Seeker available here.
```

Comment:
make an issue for this. Technically kopia returns items that are like the below so we could get a `Seeker` here if we needed. It would be quite a few changes to update the interfaces, but still possible
```go
type Reader interface {
  io.ReadCloser
  io.Seeker
}
```

Reply:
https://github.com/alcionai/corso/issues/3520